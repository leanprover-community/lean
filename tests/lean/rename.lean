example {α β} (a : α) (b : β) : unit :=
begin
  rename a a',              -- rename-compatible syntax
  guard_hyp a' := α,

  rename a' → a,            -- more suggestive syntax
  guard_hyp a := α,

  rename [a a', b b'],      -- parallel renaming
  guard_hyp a' := α,
  guard_hyp b' := β,

  rename [a' → a, b' → b],  -- ditto with alternative syntax
  guard_hyp a := α,
  guard_hyp b := β,

  rename [a → b, b → a],    -- renaming really is parallel
  guard_hyp a := β,
  guard_hyp b := α,

  rename b a,               -- shadowing is allowed (but guard_hyp doesn't like it)

  rename d e,               -- cannot rename nonexistent hypothesis
  exact ()
end


example {α} [decidable α] (a : α) : unit :=
begin
  rename a a', -- renaming works after frozen local instances

  rename α α', -- renaming doesn't work before frozen local instances
end
