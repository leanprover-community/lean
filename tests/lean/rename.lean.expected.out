rename.lean:23:2: error: Cannot rename these hypotheses:
d
This is because these hypotheses either do not occur in the
context or they occur before a frozen local instance.
In the latter case, try `tactic.unfreeze_local_instances`.
state:
α : Sort ?,
β : Sort ?,
a : α,
a : β
⊢ unit
rename.lean:32:2: error: Cannot rename these hypotheses:
α
This is because these hypotheses either do not occur in the
context or they occur before a frozen local instance.
In the latter case, try `tactic.unfreeze_local_instances`.
state:
α : Prop,
_inst_1 : decidable α,
a' : α
⊢ unit
